prov/util/src/util_ctx.c

```diff
@@ -0,0 +1,273 @@
+			"Duplicate CNTR binding\n");					\
+		return -FI_EINVAL;							\
+	}										\
+											\
+	return FI_SUCCESS;								\
+}											\
+ofi_cntr_create_setter(msg, type);							\
+ofi_cntr_create_setter(rma, type);							\
+ofi_cntr_create_setter(rem_rma, type);							\
+static int ofi_ ## type ## _ctx_bind_cntr(struct util_ctx *ctx, struct util_cntr *cntr,	\
+					  uint64_t flags)				\
+{											\
+	int ret;									\
+											\
+	ret = ofi_ ## type ## _ctx_bind_cntr(ctx, cntr, flags);				\
+	if (ret)									\
+		return ret;								\
+	ofi_ ## type ## _ctx_set_msg_cntr(ctx, cntr, flags);				\
+	ofi_ ## type ## _ctx_set_rma_cntr(ctx, cntr, flags);				\
+	ofi_ ## type ## _ctx_set_rem_rma_cntr(ctx, cntr, flags);			\
+											\
+	return FI_SUCCESS;								\
+}
+
+ofi_ctx_define(tx);
+ofi_ctx_define(rx);
+
+int ofi_ctx_bind_valid(const struct fi_provider *prov, struct util_ctx *ctx,
+		       struct fid *bfid, uint64_t flags)
+{
+	if (ctx->ep_fid.fid.fclass == FI_CLASS_RX_CTX)
+		return ofi_rx_ctx_bind_valid(prov, bfid, flags);
+	else if (ctx->ep_fid.fid.fclass == FI_CLASS_TX_CTX)
+		return ofi_tx_ctx_bind_valid(prov, bfid, flags);
+	return -FI_EINVAL;
+}
+
+int ofi_ctx_bind_cq(struct util_ctx *ctx, struct util_cq *cq, uint64_t flags)
+{
+	if (ctx->ep_fid.fid.fclass == FI_CLASS_RX_CTX)
+		return ofi_rx_ctx_bind_cq(ctx, cq, flags);
+	else if (ctx->ep_fid.fid.fclass == FI_CLASS_TX_CTX)
+		return ofi_tx_ctx_bind_cq(ctx, cq, flags);
+	return -FI_EINVAL;
+}
+
+int ofi_ctx_bind_cntr(struct util_ctx *ctx, struct util_cntr *cntr, uint64_t flags)
+{
+	if (ctx->ep_fid.fid.fclass == FI_CLASS_RX_CTX)
+		return ofi_rx_ctx_bind_cntr(ctx, cntr, flags);
+	else if (ctx->ep_fid.fid.fclass == FI_CLASS_TX_CTX)
+		return ofi_tx_ctx_bind_cntr(ctx, cntr, flags);
+	return -FI_EINVAL;
+}
+
+int ofi_ctx_bind(struct util_ctx *ctx, struct fid *fid, uint64_t flags)
+{
+	int ret;
+	struct util_cq   *cq;
+	struct util_cntr *cntr;
+
+	ret = ofi_ctx_bind_valid(ctx->domain->prov, ctx, fid, flags);
+	if (ret)
+		return ret;
+
+	switch (fid->fclass) {
+		case FI_CLASS_CQ:
+			cq = container_of(fid, struct util_cq, cq_fid.fid);
+			return ofi_ctx_bind_cq(ctx, cq, flags);
+		case FI_CLASS_CNTR:
+			cntr = container_of(fid, struct util_cntr, cntr_fid.fid);
+			return ofi_ctx_bind_cntr(ctx, cntr, flags);
+	}
+
+	return -FI_EINVAL;
+}
+
+int ofi_ctx_init(struct fid_domain *domain, const struct util_prov *util_prov,
+		 struct fi_info *info, struct util_ctx *ctx, size_t fclass,
+		 void *context, ofi_ctx_progress_func progress)
+{
+	struct util_domain *util_domain;
+	int ret;
+
+	switch (fclass) {
+	case FI_CLASS_RX_CTX:
+		ret = ofi_tx_ctx_init(info, ctx);
+		break;
+	case FI_CLASS_TX_CTX:
+		ret = ofi_tx_ctx_init(info, ctx);
+		break;
+	default:
+		ret = -FI_EINVAL;
+		break;
+	}
+
+	util_domain = container_of(domain, struct util_domain, domain_fid);
+	ret = ofi_prov_check_info(util_prov,
+				  util_domain->fabric->fabric_fid.api_version,
+				  info);
```

Comment:
I'm assuming ofi_ctx_init would get called either from a *_scalable_ep() function or a *_tx_context() function. In both cases I think this check would not be needed since it would have already been done in *_scalable_ep or *_tx_context() need not pass a fi_info arg to ofi_ctx_init.